### PR TITLE
Userland: Let's talk about ContentAccessServer

### PR DESCRIPTION
**Pull request by alimpfard**
`pp http://place/yak.png` time!

### CODE REVIEW

**Comment by LucasChollet** [comment]
I am aware of the discussion that happened on discord, but I'm definitely not fond of the code duplication here.
Especially when we consider that FSAS is in a transition period, duplicating after #17026 is merged would be better IMO. Yes, I'm looking at you `typename AnyResult`.

**Comment by alimpfard** [comment]
:thinking: CAS doesn't have the deprecated stuff though?

**Comment by LucasChollet** [comment]
That's not as bad as my comment sound, my bad. Still, `handle_promise` doesn't need to be templated.

Again, it wasn't crystal clear, but it was more as a note for why we should not create another service than a classic PR comment. On that point, I think that FSAS (Client included) isn't in a superb shape, and it would be nice to improve it (on my :yakstack:). A quick glance at this PR looks like you're duplicating the worst part :sweat:, so I'm not super excited for CAS in its current form. Nonetheless, this feature is really neat!

**Comment by alimpfard** [comment]
I'll remove the templated return type, sure.
Not sure what part of it you dislike (if you feel like it, you could point them out here, I wouldn't mind fixing them separately in CAS) but most of the changes you've made in #17026 are in here.

File: Userland/Services/ContentAccessServer/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+set(GENERATED_SOURCES
+    ContentAccessServerEndpoint.h
+    ContentAccessClientEndpoint.h
+)
+
+serenity_bin(ContentAccessServer)
+target_link_libraries(ContentAccessServer PRIVATE LibCore LibIPC LibGfx LibGUI LibMain LibFileSystemAccessClient LibProtocol)
+add_dependencies(ContentAccessServer WindowServer)
```
Comment (krkk) [inline Userland/Services/ContentAccessServer/CMakeLists.txt:22]: i think we should also add RequestServer to the list

```
28.799 SystemServer(24:24): RequestServer: binary "/bin/RequestServer" does not exist, skipping service.
28.799 SystemServer(24:24): No such file or directory (errno=2)
```

File: Userland/Services/ContentAccessServer/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+serenity_component(
+    ContentAccessServer
+    REQUIRED
```
Comment (krkk) [inline Userland/Services/ContentAccessServer/CMakeLists.txt:3]: let’s not bloat the minimal build config and add `DEPENDS ContentAccessServer` to each `serenity_component()` that requires this instead :^)

Comment (alimpfard) [reply Userland/Services/ContentAccessServer/CMakeLists.txt:3]: sure. idk why anyone would want a "minimal build", but that's irrelevant to this discussion

File: Userland/Services/ContentAccessServer/ConnectionFromClient.cpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+            label.set_text(DeprecatedString::formatted("Downloading {} from {}", human_readable_size(*total), url));
+            progressbar.set_range(0, *total);
+            progressbar.set_value(current);
+            progressbar.set_format(GUI::Progressbar::Percentage);
+        },
+    };
+}
+
+void ConnectionFromClient::request_url_handler(i32 request_id, i32 window_server_client_id, i32 parent_window_id, URL const& url, ShouldPrompt prompt)
```
Comment (konradekk) [inline Userland/Services/ContentAccessServer/ConnectionFromClient.cpp:101]: i must tell that itʼs pretty hard to follow! it seems it could be divided a little—definitely at L145 (maybe in other places, probably also at L193).

it seems that you could extract a boolean-returning function that would just return `approved` value immediately (named e.g. `is_remote_access_allowed` or so).

File: Userland/Services/ContentAccessServer/ConnectionFromClient.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+ConnectionFromClient::ProgressObject ConnectionFromClient::create_download_progress_window(i32 window_server_client_id, i32 parent_window_id)
+{
+    auto main_window = create_dummy_child_window(window_server_client_id, parent_window_id);
+    auto dialog = GUI::Window::construct(main_window);
+    dialog->set_title("Downloading Remote File..."sv);
+    dialog->set_rect(Gfx::IntRect { 0, 0, 400, 100 });
+    dialog->center_within(*main_window);
+    auto frame = MUST(dialog->set_main_widget<GUI::Widget>());
+    MUST(frame->load_from_gml(R"gmlgmlgml(
```
Comment (konradekk) [inline Userland/Services/ContentAccessServer/ConnectionFromClient.cpp:58]: i thought it maybe would be better to go with `TRY`/`ErrorOr` instead (if possible), as this is a new piece of code…? (pushing `MUST` as high as possible, that would probably L238? 🤔)

Comment (alimpfard) [reply Userland/Services/ContentAccessServer/ConnectionFromClient.cpp:58]: if we can't load the GML, there's no point in doing anything at all, might as well crash right where it failed imo.

Comment (konradekk) [reply Userland/Services/ContentAccessServer/ConnectionFromClient.cpp:58]: thereʼs something else here to be `TRIED`… but sure thing! 👍

**Comment by alimpfard** [comment]
Eh, will revisit this later maybe.
